### PR TITLE
kv: write a file based membuffer which might be used for persistented large transaction 

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

The large transaction costs a huge amount of memory.
If we can persistent the data to file, maybe we can reduce the memory.

### What is changed and how it works?

Write a `fileDBBuffer` that implements the MemBuffer.
Wrap the `lazyMemBuffer` to use it when the `MemDbBuffer` is large.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-ansible` repository

Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

File: kv/union_store.go [hunk 1]
```diff
@@ -133,14 +141,48 @@ func (lmb *lazyMemBuffer) Get(ctx context.Context, k Key) ([]byte, error) {
 	return lmb.mb.Get(ctx, k)
 }
 
+const saveToFile = 20 *1024*1024
```
Comment (qw4990) [inline kv/union_store.go]: 20MB? It seems too small.

**Comment by tiancaiamao** [comment]
Amazing! I have test up to 2G transaction on my own laptop (8G mem) with mocktikv!

During the 2pc phase, there are heavy load read/write the disk:

image

and the memory occupation is rather small, mostly used by mocktikv that we can ignore:

image

File: kv/union_store.go (hunk 1 above)
Comment (tiancaiamao) [reply kv/union_store.go]: Testing, still WIP

**Comment by XuHuaiyu** [comment]
What's the amount of memory that tidb-server will use before these optimizations?

**Comment by tiancaiamao** [comment]
> What's the amount of memory that tidb-server will use before these optimizations?

It can't run! 100M transaction is already a hard work for my laptop before this change, I guess 300M at most.

**Pull request closed without merging**